**Task 5- Credit Card Fraud Detection**

In [1]:
#Import the required libraries
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns

In [2]:
#Load the dataset
data = pd.read_csv("/content/creditcard.csv")

In [3]:
#Print some rows of data
data.head()

Time        V1        V2        V3        V4        V5        V6        V7  \
0   0.0 -1.359807 -0.072781  2.536347  1.378155 -0.338321  0.462388  0.239599   
1   0.0  1.191857  0.266151  0.166480  0.448154  0.060018 -0.082361 -0.078803   
2   1.0 -1.358354 -1.340163  1.773209  0.379780 -0.503198  1.800499  0.791461   
3   1.0 -0.966272 -0.185226  1.792993 -0.863291 -0.010309  1.247203  0.237609   
4   2.0 -1.158233  0.877737  1.548718  0.403034 -0.407193  0.095921  0.592941   

         V8        V9  ...       V21       V22       V23       V24       V25  \
0  0.098698  0.363787  ... -0.018307  0.277838 -0.110474  0.066928  0.128539   
1  0.085102 -0.255425  ... -0.225775 -0.638672  0.101288 -0.339846  0.167170   
2  0.247676 -1.514654  ...  0.247998  0.771679  0.909412 -0.689281 -0.327642   
3  0.377436 -1.387024  ... -0.108300  0.005274 -0.190321 -1.175575  0.647376   
4 -0.270533  0.817739  ... -0.009431  0.798278 -0.137458  0.141267 -0.206010   

        V26       V27       V28  Amount  Class  
0 -0.189115  0.133558 -0.021053  149.62      0  
1  0.125895 -0.008983  0.014724    2.69      0  
2 -0.139097 -0.055353 -0.059752  378.66      0  
3 -0.221929  0.062723  0.061458  123.50      0  
4  0.502292  0.219422  0.215153   69.99      0  

[5 rows x 31 columns]

In [4]:
data.shape

(284807, 31)

In [5]:
#Describe the data
data.describe()

Time            V1            V2            V3            V4  \
count  284807.000000  2.848070e+05  2.848070e+05  2.848070e+05  2.848070e+05   
mean    94813.859575  1.168375e-15  3.416908e-16 -1.379537e-15  2.074095e-15   
std     47488.145955  1.958696e+00  1.651309e+00  1.516255e+00  1.415869e+00   
min         0.000000 -5.640751e+01 -7.271573e+01 -4.832559e+01 -5.683171e+00   
25%     54201.500000 -9.203734e-01 -5.985499e-01 -8.903648e-01 -8.486401e-01   
50%     84692.000000  1.810880e-02  6.548556e-02  1.798463e-01 -1.984653e-02   
75%    139320.500000  1.315642e+00  8.037239e-01  1.027196e+00  7.433413e-01   
max    172792.000000  2.454930e+00  2.205773e+01  9.382558e+00  1.687534e+01   

                 V5            V6            V7            V8            V9  \
count  2.848070e+05  2.848070e+05  2.848070e+05  2.848070e+05  2.848070e+05   
mean   9.604066e-16  1.487313e-15 -5.556467e-16  1.213481e-16 -2.406331e-15   
std    1.380247e+00  1.332271e+00  1.237094e+00  1.194353e+00  1.098632e+00   
min   -1.137433e+02 -2.616051e+01 -4.355724e+01 -7.321672e+01 -1.343407e+01   
25%   -6.915971e-01 -7.682956e-01 -5.540759e-01 -2.086297e-01 -6.430976e-01   
50%   -5.433583e-02 -2.741871e-01  4.010308e-02  2.235804e-02 -5.142873e-02   
75%    6.119264e-01  3.985649e-01  5.704361e-01  3.273459e-01  5.971390e-01   
max    3.480167e+01  7.330163e+01  1.205895e+02  2.000721e+01  1.559499e+01   

       ...           V21           V22           V23           V24  \
count  ...  2.848070e+05  2.848070e+05  2.848070e+05  2.848070e+05   
mean   ...  1.654067e-16 -3.568593e-16  2.578648e-16  4.473266e-15   
std    ...  7.345240e-01  7.257016e-01  6.244603e-01  6.056471e-01   
min    ... -3.483038e+01 -1.093314e+01 -4.480774e+01 -2.836627e+00   
25%    ... -2.283949e-01 -5.423504e-01 -1.618463e-01 -3.545861e-01   
50%    ... -2.945017e-02  6.781943e-03 -1.119293e-02  4.097606e-02   
75%    ...  1.863772e-01  5.285536e-01  1.476421e-01  4.395266e-01   
max    ...  2.720284e+01  1.050309e+01  2.252841e+01  4.584549e+00   

                V25           V26           V27           V28         Amount  \
count  2.848070e+05  2.848070e+05  2.848070e+05  2.848070e+05  284807.000000   
mean   5.340915e-16  1.683437e-15 -3.660091e-16 -1.227390e-16      88.349619   
std    5.212781e-01  4.822270e-01  4.036325e-01  3.300833e-01     250.120109   
min   -1.029540e+01 -2.604551e+00 -2.256568e+01 -1.543008e+01       0.000000   
25%   -3.171451e-01 -3.269839e-01 -7.083953e-02 -5.295979e-02       5.600000   
50%    1.659350e-02 -5.213911e-02  1.342146e-03  1.124383e-02      22.000000   
75%    3.507156e-01  2.409522e-01  9.104512e-02  7.827995e-02      77.165000   
max    7.519589e+00  3.517346e+00  3.161220e+01  3.384781e+01   25691.160000   

               Class  
count  284807.000000  
mean        0.001727  
std         0.041527  
min         0.000000  
25%         0.000000  
50%         0.000000  
75%         0.000000  
max         1.000000  

[8 rows x 31 columns]

In [6]:
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 284807 entries, 0 to 284806
Data columns (total 31 columns):
 #   Column  Non-Null Count   Dtype  
---  ------  --------------   -----  
 0   Time    284807 non-null  float64
 1   V1      284807 non-null  float64
 2   V2      284807 non-null  float64
 3   V3      284807 non-null  float64
 4   V4      284807 non-null  float64
 5   V5      284807 non-null  float64
 6   V6      284807 non-null  float64
 7   V7      284807 non-null  float64
 8   V8      284807 non-null  float64
 9   V9      284807 non-null  float64
 10  V10     284807 non-null  float64
 11  V11     284807 non-null  float64
 12  V12     284807 non-null  float64
 13  V13     284807 non-null  float64
 14  V14     284807 non-null  float64
 15  V15     284807 non-null  float64
 16  V16     284807 non-null  float64
 17  V17     284807 non-null  float64
 18  V18     284807 non-null  float64
 19  V19     284807 non-null  float64
 20  V20     284807 non-null  float64
 21  V21     28

In [7]:
#Check for any null values in the data
data.isnull().sum()

Time      0
V1        0
V2        0
V3        0
V4        0
V5        0
V6        0
V7        0
V8        0
V9        0
V10       0
V11       0
V12       0
V13       0
V14       0
V15       0
V16       0
V17       0
V18       0
V19       0
V20       0
V21       0
V22       0
V23       0
V24       0
V25       0
V26       0
V27       0
V28       0
Amount    0
Class     0
dtype: int64

In [8]:
#Counts the values of normal and fraud transactions.
# 0 - normal
# 1 - fraud
data["Class"].value_counts()

Class
0    284315
1       492
Name: count, dtype: int64

This data is highly unbalanced.

In [9]:
#for fraud = 1
# otherwise 0
normal_data = data[data.Class == 0]
fraud_data = data[data.Class == 1]

In [10]:
print(normal_data.shape)
print(fraud_data.shape)

(284315, 31)
(492, 31)


In [11]:
normal_data.Amount.describe()

count    284315.000000
mean         88.291022
std         250.105092
min           0.000000
25%           5.650000
50%          22.000000
75%          77.050000
max       25691.160000
Name: Amount, dtype: float64

In [12]:
fraud_data.Amount.describe()

count     492.000000
mean      122.211321
std       256.683288
min         0.000000
25%         1.000000
50%         9.250000
75%       105.890000
max      2125.870000
Name: Amount, dtype: float64

Mean of fraud data is too much more than the mean of normal data.
Max value is also less for it.

In [13]:
#compare the other columns values for normal and fraud data
data.groupby("Class").mean()

Time        V1        V2        V3        V4        V5  \
Class                                                                   
0      94838.202258  0.008258 -0.006271  0.012171 -0.007860  0.005453   
1      80746.806911 -4.771948  3.623778 -7.033281  4.542029 -3.151225   

             V6        V7        V8        V9  ...       V20       V21  \
Class                                          ...                       
0      0.002419  0.009637 -0.000987  0.004467  ... -0.000644 -0.001235   
1     -1.397737 -5.568731  0.570636 -2.581123  ...  0.372319  0.713588   

            V22       V23       V24       V25       V26       V27       V28  \
Class                                                                         
0     -0.000024  0.000070  0.000182 -0.000072 -0.000089 -0.000295 -0.000131   
1      0.014049 -0.040308 -0.105130  0.041449  0.051648  0.170575  0.075667   

           Amount  
Class              
0       88.291022  
1      122.211321  

[2 rows x 30 columns]

Too much variation in the values.

Under Sampling- We take the same number of normal transactions as the fraud transactions.
normal_sample_data--->492 values out of 284315.

In [14]:
normal_sample_data = normal_data.sample(n=492)

In [15]:
#Concatenate the both data
new_data = pd.concat([normal_sample_data,fraud_data],axis=0)

In [16]:
new_data.head()

Time        V1        V2        V3        V4        V5        V6  \
100541   67544.0 -1.775191  1.333562  0.770020  1.626936 -0.797466 -0.973988   
156378  107991.0  1.975676 -0.259357 -0.178009  0.557442 -0.724839 -0.738642   
244212  152252.0 -1.193123 -0.884721  0.202267  1.552885  2.550302 -1.486148   
88310    62081.0 -1.140229  0.670730  2.073208  0.830030 -0.823862  1.067891   
61250    49730.0  1.369046 -0.260267  0.260518 -0.522746 -0.931905 -1.434973   

              V7        V8        V9  ...       V21       V22       V23  \
100541 -0.200268  0.683767 -0.340983  ...  0.067129  0.143955 -0.228042   
156378 -0.621297 -0.092948  2.792746  ... -0.272258 -0.467577  0.379567   
244212 -0.665182  0.045780 -0.559545  ... -0.113368 -0.759147  0.352912   
88310  -0.174899  0.922210 -0.372364  ...  0.194852  0.338965 -0.015763   
61250  -0.025727 -0.396865 -1.253499  ... -0.769990 -1.792356  0.313729   

             V24       V25       V26       V27       V28  Amount  Class  
100541  0.938647 -0.145530 -0.320633 -0.154823 -0.016652   15.38      0  
156378 -0.166616 -0.449806 -0.956380  0.026293 -0.035055    9.39      0  
244212  0.508120 -0.587743  0.596818  0.088154  0.233059    1.98      0  
88310  -0.334058  0.188354 -0.231557 -0.024907  0.018667  102.47      0  
61250   0.709073 -0.032816  0.599936 -0.071861  0.015230   17.98      0  

[5 rows x 31 columns]

In [17]:
new_data["Class"].value_counts()

Class
0    492
1    492
Name: count, dtype: int64

In [18]:
new_data.groupby("Class").mean()

Time        V1        V2        V3        V4        V5  \
Class                                                                   
0      93058.319106 -0.037754  0.010979  0.082739 -0.023159  0.041262   
1      80746.806911 -4.771948  3.623778 -7.033281  4.542029 -3.151225   

             V6        V7        V8        V9  ...       V20       V21  \
Class                                          ...                       
0     -0.054803  0.034983  0.034070 -0.075504  ... -0.043930  0.010961   
1     -1.397737 -5.568731  0.570636 -2.581123  ...  0.372319  0.713588   

            V22       V23       V24       V25       V26       V27       V28  \
Class                                                                         
0      0.015489  0.004936  0.050569  0.004717  0.000309 -0.009460  0.005489   
1      0.014049 -0.040308 -0.105130  0.041449  0.051648  0.170575  0.075667   

           Amount  
Class              
0       73.891524  
1      122.211321  

[2 rows x 30 columns]

In [19]:
#Spliting data into features and target
# x= features
# y= target
x = new_data.drop(columns="Class",axis=1)
y = new_data["Class"]

In [20]:
print(x)

            Time        V1        V2        V3        V4        V5        V6  \
100541   67544.0 -1.775191  1.333562  0.770020  1.626936 -0.797466 -0.973988   
156378  107991.0  1.975676 -0.259357 -0.178009  0.557442 -0.724839 -0.738642   
244212  152252.0 -1.193123 -0.884721  0.202267  1.552885  2.550302 -1.486148   
88310    62081.0 -1.140229  0.670730  2.073208  0.830030 -0.823862  1.067891   
61250    49730.0  1.369046 -0.260267  0.260518 -0.522746 -0.931905 -1.434973   
...          ...       ...       ...       ...       ...       ...       ...   
279863  169142.0 -1.927883  1.125653 -4.518331  1.749293 -1.566487 -2.010494   
280143  169347.0  1.378559  1.289381 -5.004247  1.411850  0.442581 -1.326536   
280149  169351.0 -0.676143  1.126366 -2.213700  0.468308 -1.120541 -0.003346   
281144  169966.0 -3.113832  0.585864 -5.399730  1.817092 -0.840618 -2.943548   
281674  170348.0  1.991976  0.158476 -2.583441  0.408670  1.151147 -0.096695   

              V7        V8        V9  .

In [21]:
print(y)

100541    0
156378    0
244212    0
88310     0
61250     0
         ..
279863    1
280143    1
280149    1
281144    1
281674    1
Name: Class, Length: 984, dtype: int64


In [22]:
#Spliting the data into training and testing data
from sklearn.model_selection import train_test_split
x_train,x_test,y_train,y_test = train_test_split(x,y,test_size=0.2,stratify=y,random_state=2)

Model Training

In [23]:
from sklearn.linear_model import LogisticRegression
model = LogisticRegression()
model.fit(x_train,y_train)

LogisticRegression()

In [24]:
#Accuracy on training data
from sklearn.metrics import accuracy_score
x_train_prediction = model.predict(x_train)
training_data_accuracy = accuracy_score(x_train_prediction,y_train)
print("Accuracy on training data:",training_data_accuracy)

Accuracy on training data: 0.9415501905972046


In [25]:
x_test_prediction = model.predict(x_test)
print(x_test_prediction)

[1 0 0 0 1 0 1 1 0 1 0 1 1 0 0 1 0 1 0 0 0 0 0 1 0 1 1 0 1 1 1 1 0 1 1 0 0
 0 0 1 1 0 0 1 1 1 1 1 0 0 0 0 0 0 0 0 0 1 0 0 1 1 1 1 1 0 1 0 0 1 0 0 0 0
 1 1 0 1 1 1 0 1 1 1 0 0 0 0 0 0 0 1 1 1 1 1 1 1 0 1 1 1 1 0 1 0 0 0 0 0 0
 0 0 0 0 1 1 0 0 0 0 0 0 0 1 0 0 1 0 1 1 0 1 1 0 1 0 1 0 1 0 0 1 1 0 0 1 0
 1 0 0 1 1 1 0 0 1 1 0 0 0 1 0 1 1 0 0 1 0 1 1 0 1 1 0 1 0 0 0 0 0 1 1 0 1
 0 0 0 0 0 1 1 0 1 1 1 0]


In [28]:
y_test.head(197)

247995    1
58659     0
58928     0
165894    0
15451     1
         ..
108258    1
12108     1
99506     1
74496     1
136430    0
Name: Class, Length: 197, dtype: int64

In [27]:
#Accuracy on testing data
from sklearn.metrics import accuracy_score
testing_data_accuracy = accuracy_score(x_test_prediction,y_test)
print("Accuracy on testing data:",testing_data_accuracy)

Accuracy on testing data: 0.9390862944162437


There is too much accuracy of training data and testing data.We can predict the fraud transactions easily.